src/blob_file_builder.cc

```diff
@@ -6,19 +6,132 @@ namespace titandb {
+    // Encode to take ownership of underlying string.
+    record.EncodeTo(&record_str);
+    sample_records_.emplace_back(record_str);
+    sample_str_len_ += record_str.size();
+    cached_contexts_.emplace_back(std::move(ctx));
+    if (cf_options_.blob_file_compression_options.zstd_max_train_bytes > 0 &&
+        sample_str_len_ >=
+            cf_options_.blob_file_compression_options.zstd_max_train_bytes) {
+      EnterUnbuffered(out_ctx);
+    }
+  } else if (builder_state_ == BuilderState::kUnbuffered) {
+    encoder_.EncodeRecord(record);
+    WriteEncoderData(&ctx->new_blob_index.blob_handle);
+    out_ctx->emplace_back(std::move(ctx));
+  }
+
+  // The keys added into blob files are in order.
+  // We do key range checks for both state
+  if (smallest_key_.empty()) {
+    smallest_key_.assign(record.key.data(), record.key.size());
+  }
+  assert(cf_options_.comparator->Compare(record.key, Slice(smallest_key_)) >=
+         0);
+  assert(cf_options_.comparator->Compare(record.key, Slice(largest_key_)) >= 0);
+  largest_key_.assign(record.key.data(), record.key.size());
+}
+
+void BlobFileBuilder::CacheContext(std::unique_ptr<BlobRecordContext> ctx) {
+  cached_contexts_.emplace_back(std::move(ctx));
+}
+
+void BlobFileBuilder::EnterUnbuffered(OutContexts* out_ctx) {
+  // Using collected samples to train the compression dictionary
+  // Then replay those records in memory, encode them to blob file
+  // When above things are done, transform builder state into unbuffered
+  std::string samples;
+  samples.reserve(sample_str_len_);
+  std::vector<size_t> sample_lens;
+
+  for (const auto& record_str : sample_records_) {
+    if (cf_options_.blob_file_compression_options.zstd_max_train_bytes > 0 &&
+        samples.size() >=
+            cf_options_.blob_file_compression_options.zstd_max_train_bytes) {
+      // Enough training data
+      break;
+    }
+    size_t copy_len =
+        cf_options_.blob_file_compression_options.zstd_max_train_bytes == 0
+            ? record_str.size()
+            : std::min(cf_options_.blob_file_compression_options
+                               .zstd_max_train_bytes -
+                           samples.size(),
+                       record_str.size());
+    samples.append(record_str, 0, copy_len);
+    sample_lens.emplace_back(copy_len);
+  }
+  std::string dict;
+  dict = ZSTD_TrainDictionary(
+      samples, sample_lens,
+      cf_options_.blob_file_compression_options.max_dict_bytes);
+
+  compression_dict_.reset(
+      new CompressionDict(dict, cf_options_.blob_file_compression,
+                          cf_options_.blob_file_compression_options.level));
+  encoder_.SetCompressionDict(*compression_dict_);
+
+  FlushSampleRecords(out_ctx);
+
+  builder_state_ = BuilderState::kUnbuffered;
+}
+
+void BlobFileBuilder::FlushSampleRecords(OutContexts* out_ctx) {
+  assert(cached_contexts_.size() >= sample_records_.size());
+  size_t i = 0, j = 0;
+  for (; i < sample_records_.size(); i++, j++) {
```

Comment:
what does this double loop do? it doesn't look correct.

Reply:
to skip the "cached" contexts, small KV pairs or blobindex during merge process would have chance in here. they should be returned the same as they were passed in. after the inner loop, the context should match the sample record.

Reply:
give `i` and `j` a better name? like `sample_idx` and `ctx_idx`.